### PR TITLE
Various quick fixes / features

### PR DESCRIPTION
**Pull request by fufroma**
Lot of quick fixes, most of them based on what /var/log/ampache/ampache.log warned me.

Sorry to push a mess like that.

Also, a few features.

Features:
- Add possibility to have a custom.js
- Add header to allow browser cache on waveform
- Add filename properties to song object
- Add an easy optimisation when it's the guest user
- Add log event when preventing Administrator to reset their password

Fix:
- Add missing doctype
- Handle $user is null
- Handle if there are no info for the user
- Handle bad object
- Handle if the json can't be decoded
- Force cast on some value before doing arithmetic
- Display filename and not the whole path and filename
- Partial fix of https://github.com/ampache/ampache/issues/3386 - Better, but not good

### CODE REVIEW

**Comment by fufroma** [comment]
@lachlan-00 feel free to amend or criticize those patch, do not hesitate if there are thing I've done the wrong way

**Comment by lachlan-00** [comment]
do you know what data column in the user_shout table is for? it doesn't seem to be used on any of the links in the site

I was assuming something like a timestamp but i haven't seen anything about it. might remove it entirely

File: src/Application/Api/Ajax/Handler/SongAjaxHandler.php
Comment (lachlan-00) [inline src/Application/Api/Ajax/Handler/SongAjaxHandler.php:1]: not really sure about what $shout->data actually is?

**Comment by lachlan-00** [comment]
k, pretty much done here i'll run it on my server this week and merge. i use the filename path in the song view a lot and not having the folder would freak out my mind so i'll see what performance is like adding the folder in as a property as well

File: src/Application/Api/Ajax/Handler/SongAjaxHandler.php
Comment (fufroma) [reply src/Application/Api/Ajax/Handler/SongAjaxHandler.php:1]: Neither do I. There was warning in the ampache.log, so I fixed it...

**Comment by fufroma** [comment]
> k, pretty much done here i'll run it on my server this week and merge. i use the filename path in the song view a lot and not having the folder would freak out my mind so i'll see what performance is like adding the folder in as a property as well

I saw filename when it was path + filename, and I though that web user don't need to know where the files are (except for the sysadmin).

Thank you for your patch, I'll take notes for my next one!

**Review by lachlan-00 — APPROVED** [review]

**Pull request merged**